### PR TITLE
Added a BsonCodecProvider and BsonCodec

### PR DESCRIPTION
**Pull request by rozza**
Allows simple implementations of Bson interfaces to be encoded.

JAVA-1763

### CODE REVIEW

**Comment by rozza** [comment]
The downside to the `BsonCodec` is it only supplies `BsonDocument.class`. It could be locked down to our concrete implementation classes. I didn't think it was worth it as its a low level (bottom of the stack Codec), I'm not sure there will be an issue. If it fails to encode it still will throw an error - just like it does now.

File: driver-core/src/test/unit/com/mongodb/client/model/BsonTestHelper.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+import java.nio.ByteBuffer;
+
+import static com.mongodb.MongoClientSettings.getDefaultCodecRegistry;
+
+final class BsonTestHelper {
+    private static final CodecRegistry CODEC_REGISTRY = getDefaultCodecRegistry();
+
+    @SuppressWarnings("unchecked")
+    static BsonDocument toBson(final Bson bson, final boolean direct) {
```
Comment (jyemin) [inline driver-core/src/test/unit/com/mongodb/client/model/BsonTestHelper.java]: This seems redundant to the BsonCodec spec tests.  Seems like those + either one of these conditionals would be sufficient.

File: driver-core/src/test/unit/com/mongodb/client/model/BsonTestHelper.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        if (direct) {
+            return bson.toBsonDocument(BsonDocument.class, CODEC_REGISTRY);
+        } else {
+            Codec<Bson> codec = (Codec<Bson>) CODEC_REGISTRY.get(bson.getClass());
```
Comment (jyemin) [inline driver-core/src/test/unit/com/mongodb/client/model/BsonTestHelper.java]: Can be simplified to:

        BsonDocumentWriter writer = new BsonDocumentWriter(new BsonDocument());
        ((Codec<Bson>) CODEC_REGISTRY.get(bson.getClass())).encode(writer, bson, EncoderContext.builder().build());
        return writer.getDocument();

File: bson/src/test/unit/org/bson/codecs/BsonCodecSpecification.groovy [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        writer.writeStartDocument()
+        writer.writeName('customBson')
+        codec.encode(writer, customBson, EncoderContext.builder().build())
+        writer.writeEndDocument()
+
+        then:
+        BsonDocument.parse('{a: 1, b:2}') == writer.getDocument().get('customBson')
+    }
+
+    def 'should throw CodecConfiguration exception if cannot encode  Bson'() {
```
[the remark this answers is not in the document]
Comment (rozza) [reply bson/src/test/unit/org/bson/codecs/BsonCodecSpecification.groovy]: Done

File: driver-core/src/test/unit/com/mongodb/client/model/BsonTestHelper.java (hunk 1 above)
Comment (rozza) [reply driver-core/src/test/unit/com/mongodb/client/model/BsonTestHelper.java]: I suppose, I considered the BsonCodec spec tests as unit tests, and these integration tests. I'll remove the direct flag and just use the codec for the core tests.

File: driver-core/src/test/unit/com/mongodb/client/model/BsonTestHelper.java (hunk 2 above)
Comment (rozza) [reply driver-core/src/test/unit/com/mongodb/client/model/BsonTestHelper.java]: Nice - done.

**Pull request closed without merging**